### PR TITLE
29 standardise location data

### PR DESCRIPTION
**Pull request by drkane**
Pull request for issue #29. Based heavily on copying code from the other additional data sources.

This command fetches data from https://github.com/drkane/geo-lookups on LSOA, MSOA, ward and LA. Each grant is checked in three ways to see if any of the geocodes for those areas are used. In order of priority they are:

- `beneficiaryLocation`
- `recipientOrganization -> Location`
- `recipientOrganizationLocation` (from `additional_data`)

If locations are found at any level then the subsequent levels aren't checked. It is possible to add more than one location though, if there were eg more than one beneficiaryLocation supplied.

The data that is added to additional_data allows for standardised local authority and higher geographies. For LSOA, MSOA & Ward it also adds a lat and longitude corresponding to the centre of the area.

One worry is that all three of these methods together make the process very slow - it was taking multiple minutes for me, even when I only tested it on 100ish grants. I wonder if we need to rethink the process and turn it into a few SQL commands executed in bulk.

### CODE REVIEW

File: datastore/additional_data/models.py [hunk 1]
```diff
@@ -92,3 +92,22 @@ class NSPL(models.Model):
...
+
+    AREA_TYPE = [
+        (LSOA, 'Lower Super Output Area'),
+        (MSOA, 'Middle Super Output Area'),
+        (LA, 'Local Authority'),
+        (WARD, 'Ward'),
+    ]
+
+    areacode = models.CharField(max_length=200, unique=True)
+    areatype = models.CharField(max_length=20, choices=AREA_TYPE)
```
Comment (michaelwood) [inline datastore/additional_data/models.py:112]: As I found out earlier, not to underestimate how much faster indexing on postgres can make things.

Think we could add some indexes on areacode and areatype 
https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/indexes/

https://github.com/ThreeSixtyGiving/datastore/pull/40/commits/73b2f02eaecc3a94bd9f4bc7b12b320503250e48#diff-592aa937943d3ac113983eb645744966

File: datastore/additional_data/sources/geo_lookup.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
+import csv
+import hashlib
+import io
+import json
+import zipfile
+from datetime import datetime
+
+import requests
+from django.forms.models import model_to_dict
+
```
Comment (michaelwood) [inline datastore/additional_data/sources/geo_lookup.py:5]: Not sure you're actually using model_to_dict?

Comment (drkane) [reply datastore/additional_data/sources/geo_lookup.py:5]: now removed

File: datastore/additional_data/models.py (hunk 1 above)
Comment (drkane) [reply datastore/additional_data/models.py]: good idea - indexes now added

File: datastore/additional_data/sources/geo_lookup.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        },
+    }
+
+    def __init__(self):
+        pass
+
+    def get_lookup_data(self, areatype, a, areas):
+        def clean_fields(row, a):
+            for k, v in row.items():
+                k = k.lower()
```
Comment (BibianaC) [inline datastore/additional_data/sources/geo_lookup.py:49]: Would it be possible to give a more meaningful name to `k`, `v`, `a`? Also in the other functions.

**Comment by michaelwood** [comment]
Master now has the other two additional data branches in it, so I'm going to create a new branch for this work with conflicts resolved and a new database migration to keep those in sequence.

**Pull request closed without merging**